### PR TITLE
ENH: Tolerate PDF with invalid xref pointed objects

### LINKED ISSUE #2326: Unable to find 'endstream' marker for obj starting at 13367.
Hello,

When executing this piece of code:

```python
from pypdf import PdfReader,PdfWriter
import traceback

try:
    input_pdf = PdfReader(dwnld_filepath)
    output_pdf = PdfWriter()
    image = input_pdf.pages[0]
    output_pdf.add_page(image)
    output_pdf.write(file_path)
except Exception as e:
    traceback.print_exc()


```

 here the PDF file(s) that cause the issue. 
 
[EveBest.pdf](https://github.com/py-pdf/pypdf/files/13544144/EveBest.pdf)

## Traceback

This is the complete traceback I see:

```
Traceback (most recent call last):
  File "/Users/shafeerali/Documents/Nanonets/avanto/API/test.py", line 58, in <module>
    output_pdf.add_page(image)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/_writer.py", line 418, in add_page
    return self._add_page(page, list.append, excluded_keys)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/_writer.py", line 331, in _add_page
    page = cast("PageObject", page_org.clone(self, False, excluded_keys))
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_data_structures.py", line 199, in clone
    d__._clone(self, pdf_dest, force_duplicate, ignore_fields, visited)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_data_structures.py", line 310, in _clone
    v.clone(pdf_dest, force_duplicate, ignore_fields)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_data_structures.py", line 199, in clone
    d__._clone(self, pdf_dest, force_duplicate, ignore_fields, visited)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_data_structures.py", line 310, in _clone
    v.clone(pdf_dest, force_duplicate, ignore_fields)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_data_structures.py", line 199, in clone
    d__._clone(self, pdf_dest, force_duplicate, ignore_fields, visited)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_data_structures.py", line 310, in _clone
    v.clone(pdf_dest, force_duplicate, ignore_fields)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_base.py", line 300, in clone
    obj.clone(pdf_dest, force_duplicate, ignore_fields)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_data_structures.py", line 116, in clone
    arr.append(data.clone(pdf_dest, force_duplicate, ignore_fields))
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_base.py", line 292, in clone
    obj = self.get_object()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_base.py", line 312, in get_object
    obj = self.pdf.get_object(self)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/_reader.py", line 1401, in get_object
    retval = read_object(self.stream, self)  # type: ignore
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_data_structures.py", line 1280, in read_object
    return DictionaryObject.read_from_stream(stream, pdf, forced_encoding)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_data_structures.py", line 538, in read_from_stream
    data["__streamdata__"] = read_unsized_from_steam(stream, pdf)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pypdf/generic/_data_structures.py", line 432, in read_unsized_from_steam
    raise PdfReadError(
pypdf.errors.PdfReadError: Unable to find 'endstream' marker for obj starting at 13367.
```

### PR DESCRIPTION
**Pull request by pubpub-zz**
closes #2326

### CODE REVIEW

File: pypdf/_xobj_image_helpers.py [hunk 1]
```diff
@@ -195,7 +195,13 @@ def bits2byte(data: bytes, size: Tuple[int, int], bits: int) -> bytes:
...
             if img.mode == "1":
                 # Two values ("high" and "low").
-                assert len(lookup) == 2 * nb, len(lookup)
+                expected_count = 2 * nb
+                if len(lookup) != expected_count:
+                    if len(lookup) < expected_count:
+                        raise PdfReadError(f"Not enough lookup values: Expected {expected_count}, got {len(lookup)}.")
+                    lookup = lookup[:expected_count]
+                    if not all(_value in WHITESPACES for _value in lookup[expected_count:]):
+                        raise PdfReadError(f"Too many lookup values: Expected {expected_count}, got {len(lookup)}.")
```
Comment (pubpub-zz) [inline pypdf/_xobj_image_helpers.py:204]:
```suggestion
                    if not all(_value in WHITESPACES for _value in lookup[expected_count:]):
                        raise PdfReadError(f"Too many lookup values: Expected {expected_count}, got {len(lookup)}.")
                    lookup = lookup[:expected_count]
```

Shouldn't it be in this order ?

Comment (stefan6419846) [reply pypdf/_xobj_image_helpers.py:204]: This has not yet been merged and is part of another/my PR, so I see no real value of including and discussing it here in your apparently completely unrelated PR.

**Comment by stefan6419846** [comment]
It seems like this PR does not only fix #2326, but attempts to solve #2303 as well (which is covered by #2234 as well, including a copy of #2331 itself as well) and includes a copy of #2331. I would argue that this should be split into three different clean PRs instead.

**Comment by MartinThoma** [comment]
https://github.com/py-pdf/pypdf/pull/2334 should be merged first as this pr (#2335) contains https://github.com/py-pdf/pypdf/pull/2334

**Comment by stefan6419846** [comment]
Still blocked by #2334.

**Comment by pubpub-zz** [comment]
the latest mods from #2334 (adding warnings) should be merged in

File: pypdf/_reader.py [hunk 2]
```diff
@@ -573,6 +573,25 @@ def read(self, stream: StreamType) -> None:
                     # non-zero-index is actually correct
             stream.seek(loc, 0)  # return to where it was
 
+        # remove wrong objects( not pointing to correct structures) - cf #2326
```
Comment (stefan6419846) [inline pypdf/_reader.py:576]:
```suggestion
        # remove wrong objects (not pointing to correct structures) - cf #2326
```

File: pypdf/_reader.py [hunk 3]
```diff
@@ -573,6 +573,25 @@ def read(self, stream: StreamType) -> None:
...
+                if gen == 65535:
+                    continue
+                ids = list(xref_entry.keys())
+                for id in ids:
+                    stream.seek(xref_entry[id], 0)
+                    try:
+                        self.read_object_header(stream)
+                    except ValueError:
+                        logger_warning(
+                            f"ignore wrong pointing object {id} {gen} (offset {xref_entry[id]})",
```
Comment (stefan6419846) [inline pypdf/_reader.py:589]:
```suggestion
                            f"Ignoring wrong pointing object {id} {gen} (offset {xref_entry[id]})",
```

File: tests/test_workflows.py [hunk 4]
```diff
@@ -547,7 +547,11 @@ def test_get_fields_warns(tmp_path, caplog, url, name):
         retrieved_fields = reader.get_fields(fileobj=fp)
 
     assert retrieved_fields == {}
-    assert normalize_warnings(caplog.text) == ["Object 2 0 not defined."]
+    assert normalize_warnings(caplog.text) == [
+        "ignore wrong pointing object 1 65536 (offset 0)",
```
Comment (stefan6419846) [inline tests/test_workflows.py:551]:
```suggestion
        "Ignoring wrong pointing object 1 65536 (offset 0)",
```

File: tests/test_workflows.py [hunk 5]
```diff
@@ -547,7 +547,11 @@ def test_get_fields_warns(tmp_path, caplog, url, name):
...
+        "ignore wrong pointing object 2 65536 (offset 0)",
```
Comment (stefan6419846) [inline tests/test_workflows.py:552]:
```suggestion
        "Ignoring wrong pointing object 2 65536 (offset 0)",
```

File: tests/test_reader.py [hunk 6]
```diff
@@ -1291,6 +1294,7 @@ def test_reader(caplog):
     caplog.clear()
     # first call requires some reparations...
     reader.pages[0].extract_text()
+    assert "found" in caplog.text
```
Comment (stefan6419846) [inline tests/test_reader.py:1297]: This seems to fail in CI.

**Review by stefan6419846 — APPROVED** [review]

**Pull request merged**